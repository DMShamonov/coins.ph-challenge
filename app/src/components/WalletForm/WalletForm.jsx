// @flow

import * as React from 'react';

import {
  Form,
  FormTitle,
  FormGroup,
  FormError,
} from 'components/Form';
import Input from 'components/Input';
import Button from 'components/Button';

import type { PropsType } from './types.js.flow';

export default function WalletForm({
  label,
  passphrase,
  error,
  hasError,
  isFormDisabled,
  onLabelChange,
  onPassphraseChange,
  onSubmit,
}: PropsType): React.Node {
  return (
    <Form onSubmit={onSubmit}>
      <FormTitle>Create wallet</FormTitle>
      <FormGroup>
        <Input
          name="label"
          label="Label"
          placeholder="Enter a wallet label"
          value={label}
          required
          disabled={isFormDisabled}
          onChange={onLabelChange}
        />
      </FormGroup>
      <FormGroup>
        <Input
          type="password"
          name="passphrase"
          label="Passphrase"
          placeholder="Enter a passphrase to encrypt wallet"
          value={passphrase}
          required
          disabled={isFormDisabled}
          onChange={onPassphraseChange}
        />
      </FormGroup>
      {hasError && <FormError>{error}</FormError>}
      <Button type="submit" disabled={isFormDisabled}>Create wallet</Button>
    </Form>
  );
}
