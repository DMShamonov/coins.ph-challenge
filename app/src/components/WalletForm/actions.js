import api from 'controllers/api';

import WALLET_FORM_ACTION_TYPES from 'redux/actionTypes/walletForm';
import WALLETS_ACTION_TYPES from 'redux/actionTypes/wallets';

export function createWallet() { // eslint-disable-line import/prefer-default-export
  return async function _createWallet(dispatch, getState) {
    dispatch({ type: WALLET_FORM_ACTION_TYPES.ON_START_PROCESSING });

    const { ui: { walletForm: { label, passphrase } } } = getState();

    try {
      const { wallet: { wallet } } = await api().wallets().createWalletWithKeychains({ label, passphrase, backupXpubProvider: 'keyternal' });

      dispatch({ type: WALLETS_ACTION_TYPES.ON_ADD, payload: { wallets: [wallet] } });
      dispatch({ type: WALLET_FORM_ACTION_TYPES.ON_RESET_FORM });
    } catch (e) {
      const error = e?.result?.message || 'Something went wrong. Try again, please.';

      dispatch({ type: WALLET_FORM_ACTION_TYPES.ON_SET_ERROR, error });
    }

    dispatch({ type: WALLET_FORM_ACTION_TYPES.ON_FINISH_PROCESSING });
  };
}
