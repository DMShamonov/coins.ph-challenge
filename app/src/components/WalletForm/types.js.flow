// @flow

export type PropsType = {|
  label: string,
  passphrase: string,
  error: string,
  hasError: boolean,
  isFormDisabled: boolean,
  onLabelChange: SyntheticInputEvent<HTMLInputElement> => void,
  onPassphraseChange: SyntheticInputEvent<HTMLInputElement> => void,
  onSubmit: Event => void,
|};
