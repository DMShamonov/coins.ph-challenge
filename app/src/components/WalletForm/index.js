import { connect } from 'react-redux';

import WALLET_FORM_ACTION_TYPES from 'redux/actionTypes/walletForm';

import { createWallet } from './actions';
import WalletForm from './WalletForm';

function mapStateToProps({ ui: { walletForm } }) {
  const { isProcessing, error, ...formData } = walletForm;
  const hasError = !!error;

  return {
    ...formData,
    error,
    hasError,
    isFormDisabled: isProcessing,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    onLabelChange({ target: { value: label } }) {
      dispatch({ type: WALLET_FORM_ACTION_TYPES.ON_CHANGE_LABEL, payload: { label } });
    },
    onPassphraseChange({ target: { value: passphrase } }) {
      dispatch({ type: WALLET_FORM_ACTION_TYPES.ON_CHANGE_PASSPHRASE, payload: { passphrase } });
    },
    onSubmit(e) {
      e.preventDefault();

      dispatch(createWallet());
    },
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(WalletForm);
