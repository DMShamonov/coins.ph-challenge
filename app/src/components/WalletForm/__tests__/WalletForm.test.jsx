import React from 'react';
import { shallow } from 'enzyme';

import WalletForm from '../WalletForm';

describe('`WalletForm` component should render correctly', () => {
  test('with default props', () => {
    expect(shallow((
      <WalletForm
        label="__LABEL__"
        passphrase="__PASSPHRASE__"
        error=""
        hasError={false}
        isFormDisabled={false}
        onLabelChange={jest.fn()}
        onPassphraseChange={jest.fn()}
        onSubmit={jest.fn()}
      />
    ))).toMatchSnapshot();
  });

  test('when `error` prop is set', () => {
    expect(shallow((
      <WalletForm
        label="__LABEL__"
        passphrase="__PASSPHRASE__"
        error="__ERROR__"
        hasError={false}
        isFormDisabled={false}
        onLabelChange={jest.fn()}
        onPassphraseChange={jest.fn()}
        onSubmit={jest.fn()}
      />
    ))).toMatchSnapshot();
  });

  test('when `hasError` prop is `true`', () => {
    expect(shallow((
      <WalletForm
        label="__LABEL__"
        passphrase="__PASSPHRASE__"
        error=""
        hasError
        isFormDisabled={false}
        onLabelChange={jest.fn()}
        onPassphraseChange={jest.fn()}
        onSubmit={jest.fn()}
      />
    ))).toMatchSnapshot();
  });

  test('when `error` prop is set and `hasError` prop is `true`', () => {
    expect(shallow((
      <WalletForm
        label="__LABEL__"
        passphrase="__PASSPHRASE__"
        error="__ERROR__"
        hasError
        isFormDisabled={false}
        onLabelChange={jest.fn()}
        onPassphraseChange={jest.fn()}
        onSubmit={jest.fn()}
      />
    ))).toMatchSnapshot();
  });

  test('when `isFormDisabled` prop is `true`', () => {
    expect(shallow((
      <WalletForm
        label="__LABEL__"
        passphrase="__PASSPHRASE__"
        error=""
        hasError={false}
        isFormDisabled
        onLabelChange={jest.fn()}
        onPassphraseChange={jest.fn()}
        onSubmit={jest.fn()}
      />
    ))).toMatchSnapshot();
  });
});
