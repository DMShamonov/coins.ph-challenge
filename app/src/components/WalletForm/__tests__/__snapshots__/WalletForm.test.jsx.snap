// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`WalletForm\` component should render correctly when \`error\` prop is set 1`] = `
<Form
  onSubmit={[MockFunction]}
>
  <FormTitle>
    Create wallet
  </FormTitle>
  <FormGroup>
    <Input
      label="Label"
      name="label"
      onChange={[MockFunction]}
      placeholder="Enter wallet label"
      required={true}
      value="__LABEL__"
    />
  </FormGroup>
  <FormGroup>
    <Input
      label="Passphrase"
      name="passphrase"
      onChange={[MockFunction]}
      placeholder="Enter passphrase to encrypt wallet"
      required={true}
      type="password"
      value="__PASSPHRASE__"
    />
  </FormGroup>
  <Button
    disabled={false}
    type="submit"
  >
    Create wallet
  </Button>
</Form>
`;

exports[`\`WalletForm\` component should render correctly when \`error\` prop is set and \`hasError\` prop is \`true\` 1`] = `
<Form
  onSubmit={[MockFunction]}
>
  <FormTitle>
    Create wallet
  </FormTitle>
  <FormGroup>
    <Input
      label="Label"
      name="label"
      onChange={[MockFunction]}
      placeholder="Enter wallet label"
      required={true}
      value="__LABEL__"
    />
  </FormGroup>
  <FormGroup>
    <Input
      label="Passphrase"
      name="passphrase"
      onChange={[MockFunction]}
      placeholder="Enter passphrase to encrypt wallet"
      required={true}
      type="password"
      value="__PASSPHRASE__"
    />
  </FormGroup>
  <FormError>
    __ERROR__
  </FormError>
  <Button
    disabled={false}
    type="submit"
  >
    Create wallet
  </Button>
</Form>
`;

exports[`\`WalletForm\` component should render correctly when \`hasError\` prop is \`true\` 1`] = `
<Form
  onSubmit={[MockFunction]}
>
  <FormTitle>
    Create wallet
  </FormTitle>
  <FormGroup>
    <Input
      label="Label"
      name="label"
      onChange={[MockFunction]}
      placeholder="Enter wallet label"
      required={true}
      value="__LABEL__"
    />
  </FormGroup>
  <FormGroup>
    <Input
      label="Passphrase"
      name="passphrase"
      onChange={[MockFunction]}
      placeholder="Enter passphrase to encrypt wallet"
      required={true}
      type="password"
      value="__PASSPHRASE__"
    />
  </FormGroup>
  <FormError />
  <Button
    disabled={false}
    type="submit"
  >
    Create wallet
  </Button>
</Form>
`;

exports[`\`WalletForm\` component should render correctly when \`isFormDisabled\` prop is \`true\` 1`] = `
<Form
  onSubmit={[MockFunction]}
>
  <FormTitle>
    Create wallet
  </FormTitle>
  <FormGroup>
    <Input
      label="Label"
      name="label"
      onChange={[MockFunction]}
      placeholder="Enter wallet label"
      required={true}
      value="__LABEL__"
    />
  </FormGroup>
  <FormGroup>
    <Input
      label="Passphrase"
      name="passphrase"
      onChange={[MockFunction]}
      placeholder="Enter passphrase to encrypt wallet"
      required={true}
      type="password"
      value="__PASSPHRASE__"
    />
  </FormGroup>
  <Button
    disabled={true}
    type="submit"
  >
    Create wallet
  </Button>
</Form>
`;

exports[`\`WalletForm\` component should render correctly with default props 1`] = `
<Form
  onSubmit={[MockFunction]}
>
  <FormTitle>
    Create wallet
  </FormTitle>
  <FormGroup>
    <Input
      label="Label"
      name="label"
      onChange={[MockFunction]}
      placeholder="Enter wallet label"
      required={true}
      value="__LABEL__"
    />
  </FormGroup>
  <FormGroup>
    <Input
      label="Passphrase"
      name="passphrase"
      onChange={[MockFunction]}
      placeholder="Enter passphrase to encrypt wallet"
      required={true}
      type="password"
      value="__PASSPHRASE__"
    />
  </FormGroup>
  <Button
    disabled={false}
    type="submit"
  >
    Create wallet
  </Button>
</Form>
`;
